### DISCUSSION TITLE
How to pass environment (.env) variables to frontend?

### DISCUSSION THREAD

**Discussion by NikolaGrgic** [tauri-apps/tauri#2873] [Q&A]
Hey all,

I have a dynamically-generated variable created in the Rust backend that I need to pass down as an environment variable to the React frontend. Is there any way to pass variables to create-react-app before Tauri launches the application?

Thanks in advance.

**Comment by m3rashid** [comment]
Also, With all respect, is there any way to change environment variables at runtime (running the build) in react-tauri ? 

Thanks in advance

**Reply by alchaplinsky** [reply]
You could just create `.env.development` and  `.env.production` files and add different values for your env variables for `development` and `production` environments.

**Reply by m3rashid** [reply]
But, that cannot be changed once the app is running. Actually, my app needs to contact a backend api which can be any IP (can change overtime). This was my usecase

I just accessed the host system environment variables (`.bashrc` or `.zshrc` (linux) and environment variables (windows)) from the rust side and invoked it on the frontend, for my part, this is resolved now .

Now, i can change the environment variables on the host and the new IP is where the api server is called 

This may not be the best approach according to tauri, but I couldn't think of a better one

**Comment by alchaplinsky** [comment]
@NikolaGrgic You should be able to do that by naming your env variable with `VITE_` prefix. This way Vite will pass these env vars to your js bundle. 

So for instance you can create a `VITE_CLIENT_ID` variable and then reference this variable as `VITE_CLIENT_ID` in your JS code.

**Comment by m3rashid** [comment]
I just implemented a similar thing in my react + tauri application.
let me show you how I did


This is my `main.rs` file
```rs
fn main() {
  tauri::Builder::default()
    .invoke_handler(tauri::generate_handler![get_environment_variable])
    .run(tauri::generate_context!())
    .expect("error while running tauri application");
}
```

Here, `get_environment_variable` is the function i want to run in the frontend, and this is its implementation

```rs
#[tauri::command]
fn get_environment_variable (name: &str) -> String {
  std::env::var(name).unwrap_or_else(|_| "".to_string())
}
```

For your use case, you can expose a similar function with no parameters which just returns your "dynamically generated variable"

```js
invoke("get_environment_variable", { name: "params" })
.then((hostIp) => {
  if (hostIp) serverRootUrl = "http://" + hostIp + ":5000";
  else serverRootUrl = defaultServerUrl;
})
.catch(console.log);
```

My case was similar because if the server URL is not loaded, there is no use of launching the app. So, I waited for a while using `setTimeout` (and in the meantime, I expect all the fetching is complete from the rust side of things), then launching the app.

Looks a bit hacky, but it is what I did
Open to suggestions, please let me know if there are better steps

**Reply by darcyvoutt** [reply]
@m3rashid Has this worked for you at build time? For me, it's only working during runtime but not during build time. I specifically only want to use the variable within Rust code, I won't need to pass it to the frontend or have a variable I don't want to store in the codebase. But even when testing this solution, it doesn't show up at build time.

**Reply by jackadair** [reply]
@darcyvoutt Going through the same thing right now, curious if you have found a solution?

**Reply by darcyvoutt** [reply]
@jackadair Thanks for poking me on this cause I forgot to post back my findings!

What I did and got advice within the Tauri Discord _(recommended), was to do the following.

```rust
fn get_secret() -> String {
    let secret = env!("SECRET");
    format!("{}", secret)
}
```

Now the Rust analyzer flags the use of `env!` but it doesn't break the building of the application. I'm a novice regarding Rust, but as I understand it, that will force the `env` to work regardless of the compiled runtime. 

Now I should flag, the secret I'm using it is meant to be a hidden string that I didn't want in the repo but isn't sensitive data (ie. personal information). It could still be found within the binary as I understand it if someone is stealthy enough, so keep in that in mind, that it's being used correctly.

**Reply by jackadair** [reply]
@darcyvoutt Wow thank you so much, as a Rust novice myself been struggling with this for a few days. Cheers my friend 🍺

**Comment by developermindsetcom** [comment]
Just came across this recently, here're several ways I found:

**1. Using `std::env`**

```rust
#[tauri::command]
fn get_env(name: &str) -> String {
    std::env::var(String::from(name)).unwrap_or(String::from(""))
}
```

Then in typescript:

```typescript
import { invoke } from "@tauri-apps/api/core";

function get_env(name: string) {
  return await invoke("get_env", { name });
}
```

**2. Using Vite**

Only with `VITE_` prefixes:

```javascript
import.meta.env.VITE_DEVELOPER_MINDSET
```

**3. Using SvelteKit**

In SSG mode only and only starting with `PUBLIC_` prefixes.

```javascript
import * as env from "$env/static/public";
console.log(import.meta.env.PUBLIC_DEVELOPER_MINDSET); // works
```

**4. Pure Rust**

Install crate: `cargo add dotenv`

Then in your rust code, load from `.env` using:

```rust
dotenv::dotenv().ok();
```

Or using macro:

```rust
#[macro_use]
extern crate dotenv_codegen;
```

And then in rust you can use `dotenv` macro:

```rust
println!("{}", dotenv!("DEVELOPER_MINDSET"));
```

There are couple more options on how to connect pulling and embedded environment variables inside backend and frontend of the Tauri app in my [tutorial](https://www.developermindset.com/tauri-env/).

**Reply by kennardpeters** [reply]
On mobile (iOS for me) the macro solution (using `dotenv_codegen`) was the only one that worked for me.

**Reply by developermindsetcom** [reply]
yes, because for mobile platforms (iOS and Android) you actually compile and move binaries to be executed on another platform (not your host platform where you compiled them), so only the code that was embedded in the binaries – stays.

**Comment by Sinet2000** [comment]
## 1. Create environment files in the project root
- `.env.staging`
- `.env.production`

## 2. Update `package.json`
Add a script to run the desktop app in staging:
```json
"dev:staging": "APP_ENV=staging tauri dev"
```

## 3. Add dotenvy dependency

In src-tauri/Cargo.toml:
```
dotenvy = "0.15"
```

## 4. Load env files in the Tauri backend

In src-tauri/src/lib.rs:
```rs
use dotenvy::from_filename;
use std::env;

fn resolve_mode() -> String {
    let mut args = env::args().peekable();
    while let Some(arg) = args.next() {
        if arg == "--env" {
            if let Some(val) = args.next() {
                return val;
            }
        }
        if let Some(mode) = arg.strip_prefix("--env=") {
            return mode.to_string();
        }
    }

    env::var("APP_ENV").unwrap_or_else(|_| "development".to_string())
}

fn load_env_files() {
    let mode = resolve_mode();
    let mode_file = format!(".env.{}", mode);

    // Try mode-specific env first
    let _ = from_filename(&mode_file);

    // Fallback to base .env
    let _ = from_filename(".env");

    if cfg!(debug_assertions) {
        eprintln!(
            "[env] mode={} (tried {}, then .env)",
            mode, mode_file
        );
    }
}

#[tauri::command]
fn get_env(name: &str) -> String {
    env::var(name).unwrap_or_default()
}

#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    load_env_files();

    tauri::Builder::default()
        .plugin(tauri_plugin_opener::init())
        .invoke_handler(tauri::generate_handler![get_env])
        .run(tauri::generate_context!())
        .expect("error while running tauri application");
}
```

## 5. Read env values in the frontend (TypeScript)
```typescript
import { useEffect } from "react";
import { invoke } from "@tauri-apps/api/core";

useEffect(() => {
  async function loadEnv() {
    setMode(await invoke<string>("get_env", { name: "APP_ENV" }));
    setLocalApiUrl(
      await invoke<string>("get_env", {
        name: "API_URL",
      })
    );
  }

  loadEnv();
}, []);
```
